keps/sig-apps/3973-consider-terminating-pods-deployment/README.md

```diff
@@ -0,0 +1,812 @@
+- Add a test wrapper for any relevant tests, to ensure that they are run with all possible PodReplacementPolicy values correctly.
+  The relevant tests are those that expect some behavior on Pod deletion, and are affected by this change.
+- Add new tests that observe rollout and scaling transitions for all possible PodReplacementPolicy values and
+  ensure that `.status.terminatingReplicas` is correctly counted when the DeploymentPodReplacementPolicy feature gate is enabled.
+
+<!--
+- <test>: <link to test coverage>
+-->
+
+##### e2e tests
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+
+We expect no non-infra related flakes in the last month as a GA graduation criteria.
+-->
+
+<!--
+test/e2e/apps/replica_set.go
+test/e2e/upgrades/apps/replica_sets.go
+
+test/e2e/apps/deployment.go
+test/e2e/upgrades/apps/deployments.go
+test/e2e/storage/utils/deployment.go
+-->
+
+- Update all existing e2e tests to account for `.status.TerminatingReplicas` field in both ReplicaSets and Deployments.
+- Test that a Deployment with `RollingUpdate` strategy and a `TerminationComplete` PodReplacementPolicy does not
+  exceed the amount of pods specified by `spec.replicas + .spec.strategy.rollingUpdate.maxSurge` when
+  rolling out new revisions and/or scaling the deployment at any point in time.
+
+<!--
+- <test>: <link to test coverage>
+-->
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature gate disabled by default.
+- Unit, enablement/disablement, e2e, and integration tests implemented and passing.
+
+
+#### Beta
+- Feature gate enabled by default.
+- Any test that checks Deployment and Replicaset status is updated to count updates to `.status.TerminatingReplicas`.
+- `.spec.podReplacementPolicy` is nil by default and preserves the original behavior.
+- E2e and integration tests are in Testgrid and linked in the KEP.
+- add new metrics to `kube-state-metrics`
+
+
+#### GA
+- Every bug report is fixed.
+- Confirm the stability of e2e and integration tests.
+- DeploymentPodReplacementPolicy feature gate is ignored.
+
+### Upgrade / Downgrade Strategy
+
+No changes required for existing cluster to use the enhancement.
+
+### Version Skew Strategy
+
+We need to consider the version skew between kube-controller-manager and the apiserver.
+
+If the feature is enabled on the apiserver, but not in the kube-controller-manager, then the `.spec.podReplacementPolicy`
+field can be set, but the feature will not function.
+
+
+If the feature is not enabled on the apiserver, and it is enabled in the kube-controller-manager, then
+- The feature cannot be used for new workloads.
+- Workloads that use this feature will revert to their default behavior. This might effect existing workloads as
```

Comment:
If FG in kube-controller-manager is enabled, this isn't true right?
The controller will continue to use the new behavior...

Reply:
I was thinking about an apiserver that does not support the field. But yes it will function if supports the field.